### PR TITLE
feat(#343): Migrate Create Program AI wizard to Embedded Agent + MCP

### LINKED ISSUE #343: Migrate post-onboarding Create Program AI wizard to Embedded Agent + MCP
> **Sequencing:** Companion to #342. Land in parallel; the final `generate-program` Edge function deletion lives in #342 since Quick Workout is the last caller standing. Decisions on shared surfaces (thread `purpose`, system prompt branching) should be made jointly across both tickets.

## Context

#295 / #338 only migrated the AI program flow **inside onboarding**. The same wizard exists post-onboarding at `/library/programs/create` (path "AI") and is still on the legacy closed-loop, fully bypassing MCP:

- **UI**: `file:src/pages/CreateProgramPage.tsx` → `file:src/components/create-program/AIGeneratingStep.tsx` + `file:src/components/create-program/AIProgramPreviewStep.tsx`
- **Hook**: `file:src/hooks/useAIGenerateProgram.ts`
- **Edge function**: `file:supabase/functions/generate-program/`
- **Persistence**: `AIProgramPreviewStep` does raw `programs` + `workout_days` + `workout_exercises` inserts. **No MCP, no commit gate.**

Net result today: a user creating their **first** program (onboarding) gets the new Embedded Agent + explicit commit gate; the **second** program (library) gets the old one-shot path — no chat, no qualitative input, no MCP. Same goal, two flows, two write paths.

## Why now

- UX inconsistency: same intent, two completely different feels depending on entry point. New users see the polished flow; engaged users (= the ones creating *more* programs) get the legacy one. Backwards.
- Code duplication: `file:src/components/create-program/AIGeneratingStep.tsx` is a near-clone of the onboarding `AIGeneratingStep` we just deleted, on a separate i18n namespace + hook we'd otherwise keep alive forever.
- Coupled cleanup: alongside the Quick Workout migration (#342), this is what makes `generate-program` deletable. Best landed in close succession.

## Framing: this is continuous coaching, not "create program v2"

The single most underrated thing in this migration: **post-onboarding program creation is fundamentally not the same product as onboarding program creation.** It is *continuous coaching*. By the time the user lands here, they already have:

- a profile,
- a (likely active) program,
- weeks or months of completed workouts,
- exercise-level progression history, plateaus, achievements.

And — critically — they have **a reason** to want a new program. The set of reasons is broad and product-meaningful:

- variety / boredom,
- plateau on N exercises (ceiling hit),
- injury or medical constraint,
- shifting priority (cut → strength, hypertrophy → endurance, etc.),
- new equipment / new schedule,
- coming back from a break.

A good agent response in this context **cannot** be derived from the onboarding form alone — it needs both **what the user has done so far** *and* **why / in what direction they want to change**. Eliciting that "why" is half the value of having an agent on this surface at all.

This reframes the whole migration: it is **not a UX cleanup**, it is the introduction of a continuous-coaching surface that makes use of the read-only MCP tools we already shipped (`getWorkoutHistory`, `getTrainingStats`, `getProgramDetails`, `listPrograms`). The Embedded Agent here should *read before it drafts*, and the system prompt should *probe before it commits to a shape*.

## Scope (draft — needs Tech Plan)

**In scope:**
- Replace the AI branch of `CreateProgramPage` (`ai-constraints` + `ai-generating` + `ai-preview` steps) with the Embedded Agent flow — likely reuse `EmbeddedAgentChatStep` + `EmbeddedAgentPreviewStep` parameterized by entry context, **but with a different system prompt + a real read-tool budget**: this surface needs the agent to read training history before drafting.
- Define the **change-motivation elicitation**: the system prompt must explicitly probe *why* the user wants a new program (variety / plateau / injury / priority shift / equipment / return from break) and *in what direction* before drafting. Default to qualitative chat if no clear motivation can be inferred.
- Reuse the existing MCP commit gate (`create_program` `dry_run: true` → user confirms → `dry_run: false`).
- Settle the thread model: today `embedded_agent_threads` enforces "at most one open thread per user" (partial unique index). For repeat program creation that constraint is wrong — user already has a `committed` thread.
- Decommission `useAIGenerateProgram`, `AIGeneratingStep` (create-program), `AIProgramPreviewStep`.

**Out of scope:**
- Template + Blank paths in `CreateProgramPage` (stay as-is, same as onboarding).
- Deletion of `file:supabase/functions/generate-program/` itself — gated on Quick Workout migration (#342) since they still share the function.
- Multiple active programs / program library reorg / new product capabilities.

## Open questions

- **System prompt — top-tier decision.** The agent needs to (a) read training history via existing MCP tools (`getWorkoutHistory`, `getTrainingStats`, `getProgramDetails`, `listPrograms`), (b) elicit the user's *change motivation*, (c) only then draft. This is structurally different from the onboarding prompt — likely **branch `prompt.ts`** rather than parameterize a single one. Worth aligning with the bilan mensuel surface (#282), which mines the same signals.
- **Read-tool budget**: how many tool calls before the agent proposes a draft? Onboarding does zero (no history). This surface probably needs 2–4. Affects latency and cost — call out in the Tech Plan.
- **Thread model**: relax the partial unique index to allow one open thread per `(user, purpose)`, or version threads by `purpose: "onboarding" | "additional_program"`? Impacts schema + `useEmbeddedAgentThread` resume logic. Decision shared with #342.
- **Profile source**: onboarding had a fresh `Onboarding form` payload in the request. Library entry only has the existing profile in DB — same shape, but where does *"I want a deload-friendly cycle this time"* go? In the chat, captured via change-motivation elicitation. Confirm in Tech Plan.
- **Analytics**: reuse `embedded_agent_*` events with a `source: "library"` payload, or fork events? Affects funnel comparability.
- **Active program switch UX**: today `useGenerateProgram` deactivates the previous active program before inserting the new one. With MCP `create_program`, who owns that side effect? Tool argument? Separate `set_active_program` (#290)?

## References

- Parent epic (closed): #295
- T123 cutover doc: `file:docs/T123_—_Cutover_+_Legacy_AI_Onboarding_Cleanup.md`
- Embedded Agent flow: `file:src/components/onboarding/EmbeddedAgentChatStep.tsx`, `file:supabase/functions/embedded-agent/`
- Thread schema: `file:supabase/migrations/20260508155713_create_embedded_agent_threads.sql`
- MCP `create_program`: `file:supabase/functions/mcp/tools/createProgram.ts`
- MCP read tools: `file:supabase/functions/mcp/tools/getWorkoutHistory.ts`, `file:supabase/functions/mcp/tools/getTrainingStats.ts`, `file:supabase/functions/mcp/tools/getProgramDetails.ts`, `file:supabase/functions/mcp/tools/listPrograms.ts`
- Companion issue: #342
- Related: #290 (`set_active_program` tool), #282 (bilan mensuel — shares the read signals)
- Glossary: `Embedded Agent thread lifecycle`, `Onboarding program commit gate` in `file:docs/CONTEXT.md`

### PR DESCRIPTION
**Pull request by PierreTsia**
## What

Migrates the **post-onboarding AI program creation wizard** at `/library/programs/create` (path "AI") from the legacy closed-loop `generate-program` + raw-Supabase inserts to the **Embedded Agent** flow, gated by the `create_program` MCP commit pattern.

Returning users now get the same coaching surface as onboarding users — adapted to the fact that they already have a profile, an active **Program**, and weeks-to-months of training history. The agent reads that history before drafting and elicits a structured **change motivation** (variety / plateau / injury / priority shift / equipment change / return from break / other) before proposing a program. **`AIConstraintStep` is removed** from this branch of the wizard.

Closes #343.

## Why

Before this PR, a user creating their **first** program (onboarding) got the polished Embedded Agent + explicit commit gate; the **second** program (library) got the legacy one-shot path — no chat, no qualitative input, no MCP. **The most engaged users — who create the most programs — got the least-evolved experience.**

This surface is fundamentally not "create program v2". By the time the user lands here, they have a reason to want a new program. Eliciting that *why* — and reading the history before drafting — is half the value of having an agent on this surface at all.

Coupled cleanup: `useAIGenerateProgram` + `AIGeneratingStep` + `AIProgramPreviewStep` + `AIConstraintStep` (under `create-program/`) are near-clones of the onboarding wizard replaced in #295. Their deletion in this PR also unblocks the eventual deletion of `supabase/functions/generate-program/` (#342's closing act).

## How

Implementation split across 8 tickets (T130–T137). Architecture locked in `docs/adr/0003-additional-program-creation-shape.md` and `docs/adr/0004-embedded-agent-thread-purpose-column.md`; detailed design in `docs/Tech_Plan_—_Migrate_Create_Program_AI_to_Embedded_Agent_#343.md`.

### Ticket map

| Ticket | Shape | Commit |
|---|---|---|
| **T130** | Extract shared program-draft helpers to `_shared/programDraft.ts` + `_shared/programGemini.ts` (decouples `embedded-agent` from `generate-program` ahead of #342's deletion). | `82f3e6d` |
| **T131** | Schema migration (5 new columns + relaxed partial unique index), thread-store rewire keyed on `(user, purpose)`, client hook API + cache key by purpose, `embedded_draft` cap bumped 3 → 10. | `31e87b4` |
| **T132** | Prompt module split: `prompt/{shared,onboarding,additional-program,index}.ts`. Per-flow ready-signal validators (motivation vocab + `constraint_overrides` bounds). Purpose-aware dispatch via discriminated unions. | `68a9c19` |
| **T133** | `lib/bundle.ts` builder (profile + active program summary + 28d training stats), 8 KB size guard (`BundleSizeExceeded`), captured once at `/open` and snapshotted into `embedded_agent_threads.bundle_context`. | `80df91a` |
| **T134** | Handler wiring for `additional_program`: `/send` validator + first-accept motivation persistence + race-free `pending_constraint_overrides`, `/draft` consumes + clears overrides, `/reject` clears overrides. Observability hooks. | `d2a8cdc` |
| **T135** | Relocate `EmbeddedAgent{Chat,Generating,Preview}Step` to `src/components/embedded-agent/` + parameterize with `purpose` + `i18nNamespace`. `OnboardingPage` rewired to pass `purpose="onboarding"`. | `de7180a` |
| **T136** | Cut over `CreateProgramPage` to the new components with `purpose="additional_program"`. Delete legacy AI branch (component + hook + schema + types). Extend analytics events with `purpose`. Add `create-program` i18n. Mocked Playwright E2E. | `4f69c86` |
| **T137** | Ops runbook (`docs/Runbook_—_Additional_Program_Creation_Flow.md`): rollback, stuck-thread investigation, validator-rejection SQL, per-purpose quota cost queries. | `1392bf1` |

Follow-up commits address PR review feedback (E2E mock state, byte-count guard, ADR completeness, quota constant mismatch).

### Locked architectural decisions (full reasoning in ADR 0003 + 0004)

| Concern | Decision |
|---|---|
| Surface shape | Pure chat — no constraints form on the AI branch |
| Agent context | Pre-loaded bundle at thread open (profile + active program summary + 4w training stats), persisted in `bundle_context` JSONB. **Zero** in-conversation MCP tool calls in v1. |
| Change motivation | Hard gate via per-flow ready-signal validator. Controlled vocab persisted on the thread row. |
| Ready signal | Existing `READY_FOR_PROGRAM_DRAFT` line carries `motivation` for this flow: `{"ready":true,"summary":"...","motivation":"plateau"}` |
| System prompt | Folder split `prompt/{shared,onboarding,additional-program,index}.ts`. Shared: locale, signal parser, brand rules. Per-flow: scope rules, user-context builder, signal schema validator. |
| Thread schema | `embedded_agent_threads` gains `purpose`, `change_motivation`, `bundle_context`, `validator_rejection_count`, `pending_constraint_overrides`. Partial unique index relaxed from `(user_id)` → `(user_id, purpose)`. |
| Quota | Share `embedded_chat` (40, unchanged) and `embedded_draft` (3 → 10) lanes with onboarding. |
| Component reuse | Relocate `EmbeddedAgent{Chat,Preview,Generating}Step` to `src/components/embedded-agent/`; parameterized via `purpose` + `i18nNamespace`. |
| Analytics | Reuse `embedded_agent_*` event names + add `purpose` payload field. `motivation` included in `embedded_agent_preview_committed`. New events: `embedded_agent_preview_committed`, `embedded_agent_motivation_classification_failed`. |
| Active program switch | No new work — `create_program` MCP tool already deactivates other active programs atomically. |
| Post-commit nav | `/library/programs` (parity with current `AIProgramPreviewStep` behavior). |

## Scope (out)

- Deletion of `supabase/functions/generate-program/` — owned by #342.
- In-conversation MCP read tools — v1 uses pre-loaded bundle only.
- "Skip the chat" UI escape — deferred to post-launch metrics.
- Modifying the active program in place / multiple active programs.
- `set_active_program` tool (#290).
- Motivation classification eval suite — post-launch follow-up.

## Test plan

- [x] Deno unit tests for `_shared/programDraft`, `embedded-agent/prompt`, `embedded-agent/lib/bundle`, `embedded-agent/threadStore`, `embedded-agent/handler`, `embedded-agent/quota`
- [x] Vitest unit tests for `EmbeddedAgent{Chat,Generating,Preview}Step` (purpose-aware rendering, analytics payloads, bundle chip)
- [x] Playwright E2E `e2e/create-program-ai.spec.ts` (path-choice → chat → motivation → ready signal → draft → preview → commit) with Gemini and MCP mocked
- [x] Onboarding regression remains green: `e2e/onboarding-embedded-agent.spec.ts` untouched
- [x] `tsc --noEmit` clean

## Rollback

See `docs/Runbook_—_Additional_Program_Creation_Flow.md` §"Rollback Procedure". Migration is forward-only by design; rollback is a revert + redeploy. RLS/quotas/retention stay intact regardless of which flows are live.

### CODE REVIEW

File: supabase/functions/embedded-agent/threadStore.ts [hunk 1]
```diff
@@ -249,6 +298,110 @@ export async function setLastPreview(
...
+  supabase: SupabaseLike,
+  thread: Thread,
+  nowIso: string = new Date().toISOString(),
+): Promise<void> {
+  const { error } = await supabase
+    .from(THREADS_TABLE)
+    .update({
+      validator_rejection_count: (thread.validator_rejection_count ?? 0) + 1,
+      updated_at: nowIso,
+    })
```
[the remark this answers is not in the document]
Comment (PierreTsia) [reply supabase/functions/embedded-agent/threadStore.ts:341]: Deferring this one on purpose — captured the trade-off in ADR 0004 §4 (`33f4d09`) so it's on record as accepted rather than overlooked.

Reasoning: `validator_rejection_count` is a prompt-side tripwire (drives "the agent should give up gracefully after N rejections"), not a billing or quota counter. A dropped increment under a multi-tab race means the agent is at most one rejection more patient than intended — a UX rounding error, not a correctness bug. The fix cost is non-trivial in this codebase (Supabase JS `.update()` doesn't expose raw SQL fragments, so atomic `col + 1` needs a new Postgres function + RPC migration + tests), and the column's behavior doesn't have the kind of edges where off-by-one matters.

If the counter ever gains a sharper behavior (billing, hard-cap, anti-abuse), the pattern should be promoted to an SQL-side increment. Tracking that as a "if/when" in the ADR rather than scope-creeping this PR.

File: supabase/functions/embedded-agent/lib/bundle.ts [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+    profile: projectProfile(profileResult.value),
+    active_program: activeProgramResult.value
+      ? projectActiveProgram(activeProgramResult.value)
+      : null,
+    recent_stats: projectRecentStats(recentStatsResult.value),
+  }
+
+  const bytes = JSON.stringify(bundle).length
+  if (bytes > BUNDLE_MAX_BYTES) {
+    throw new BundleSizeExceeded(bytes)
```
[the remark this answers is not in the document]
Comment (PierreTsia) [reply supabase/functions/embedded-agent/lib/bundle.ts]: Addressed in `ba1d427` (ba1d427a4ba15ebf9f9581a149823166cf1e93e1) — switched the size guard to `new TextEncoder().encode(json).length`. Added a regression test using 4500 × `é` (4500 code units / 9000 UTF-8 bytes) to lock the behavior.

File: docs/adr/0004-embedded-agent-thread-purpose-column.md [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+- A captured-at-thread-open **bundle context** (profile + active program + 4-week stats), referenced by every assistant turn.
+- A structured **change motivation** (controlled vocabulary), required for the additional-program flow's ready-signal gate, NULL for onboarding.
+- A way to distinguish *which flow* a thread belongs to — for resume logic, analytics joins, future flow registration (#282 bilan mensuel will be #3).
+
+This is a schema change on a production table with RLS policies, a partial unique index, and a quota system that reads from it. It's hard to walk back once data flows in. It deserves to exist as a checkpoint.
+
+## Decision
+
+We will extend `embedded_agent_threads` with three new columns and relax the partial unique index in a single migration.
```
[the remark this answers is not in the document]
Comment (PierreTsia) [reply docs/adr/0004-embedded-agent-thread-purpose-column.md]: Addressed in `33f4d09` (33f4d09fd1589c0e275b3e3740dc0ac0e7cbac3b) — ADR 0004's Decision section now documents all 5 columns (added `validator_rejection_count` §4 and `pending_constraint_overrides` §5; renumbered the index swap to §6 and backfill to §7).

File: docs/CONTEXT.md [hunk 4]
```diff
@@ -47,7 +47,7 @@ The structured onboarding questionnaire in the PWA. **V1:** keep it — do not r
 → `file:src/pages/OnboardingPage.tsx`
 
 **Embedded Agent onboarding (v1)**:
-The **Embedded Agent** runs **only** after **PathChoice** when the user picks the **AI program** path. **Template** and **Blank** paths **do not** include the chat — keep scope tight.
+The **Onboarding program creation flow** (first program, no prior history) runs the **Embedded Agent** **only** after **PathChoice** when the user picks the **AI program** path. **Template** and **Blank** paths **do not** include the chat — keep scope tight. **Naming note:** historically the term collapsed channel (**Embedded Agent**) and flow (onboarding's program creation) into one label; a sibling flow now exists — see **Additional program creation flow** — sharing the same channel, not the same product purpose.
```
[the remark this answers is not in the document]
Comment (PierreTsia) [reply docs/CONTEXT.md:50]: Addressed — PR description rewritten to reflect the full T130–T137 implementation scope (ticket-by-ticket map, locked decisions, test plan, rollback pointer). No commit needed; `gh pr edit 350 --body` only.

File: supabase/functions/_shared/aiQuota.ts [hunk 5]
```diff
@@ -18,17 +18,23 @@ const QUOTA_WHITELISTED = 5
...
+ * ≈ ~9/year peak demand. 3/24h saturates fast for repeat creators;
+ * 10/24h leaves headroom while staying bounded. See ADR 0003.
  */
 export const QUOTA_REGULAR_BY_SOURCE: Record<AIGenerationSource, number> = {
   program: 5,
   workout: 5,
   embedded_chat: 40,
-  embedded_draft: 3,
+  embedded_draft: 10,
   quick_workout: 10,
```
[the remark this answers is not in the document]
Comment (PierreTsia) [reply supabase/functions/_shared/aiQuota.ts:38]: Addressed in `c6cdb92` (c6cdb92eadf335ba982693acc525e6944fcc82d7) — bumped `EMBEDDED_DRAFTS_PER_24H` from 3 to 10 to match `QUOTA_REGULAR_BY_SOURCE.embedded_draft`. Also dropped the hardcoded "3" from `onboarding.draftQuotaBody` (en + fr) so the copy now follows the cap without further edits — the create-program copy was already number-free.
